src/bika/lims/browser/worksheet/tools.py

```diff
@@ -93,3 +101,25 @@ def getWorksheetLayouts():
         [layouts.append(layout) for layout in layout_utility.getLayouts()]
 
     return DisplayList(tuple(layouts))
+
+
+def get_all_worksheet_print_templates():
+    """ Getting worksheet templates for printing
+    """
+    all_tpl = []
+    directories = sorted(iterDirectoriesOfType(WS_TEMPLATES_ADDON_DIR), key=lambda d: d.__name__)
+    for templates_resource in directories:
+        prefix = templates_resource.__name__
+        templates = [tpl for tpl in templates_resource.listDirectory() if tpl.endswith('.pt')]
+        for template in sorted(templates):
+            all_tpl.append('{0}:{1}'.format(prefix, template))
+
+    order = get_registry_record(WS_TEMPLATES_REGISTRY_NAME, default=[]) or []
+    ordered_templates = list(
+        filter(lambda item: item,
+               sorted(all_tpl, key=lambda item: order.index(item) if item in order else len(order))))
+
+    if not (sorted(ordered_templates) == sorted(order)):
+        set_registry_record(WS_TEMPLATES_REGISTRY_NAME, ordered_templates)
```

Comment:
oh, yes, of course, but i'm little confused (
Will be right use: `api.portal.set_registry_record` from `plone.api`?

Reply:
It is in general not good to write into the database when calling a getter.
Therefore, it is a better approach to use the `defaultFactory` of the schema field to return (not set) the default values to set on the field. E.g. like here:

https://github.com/senaite/senaite.core/blob/2.x/src/senaite/core/content/senaitesetup.py#L69

You would use then `get_registry_record`. However, take care if the upgrade step hasn't run yet. So you probably need to wrap it into a `try ... except` block.

Reply:
Apologies. You probably just need to return the `ordered_templates` from above, because the factory is only called if the field is empty. So there is no registry lookup required.

Reply:
This condition with checking ordering in the current templates and stored order in the registry. And auto saved if has not equal actual set templates.

Reply:
Yes, please refactor the code to something like this (warning: untested code below):

```python
def get_all_print_templates():
    """Returns the default WS print templates
    """
    print_templates = []
    directories = sorted(iterDirectoriesOfType(WS_TEMPLATES_ADDON_DIR), key=lambda d: d.__name__)
    for templates_resource in directories:
        prefix = templates_resource.__name__
        templates = [tpl for tpl in templates_resource.listDirectory() if tpl.endswith('.pt')]
        for template in sorted(templates):
            print_templates.append('{0}:{1}'.format(prefix, template))
    return print_templates


def get_print_template_order():
    """Return the print template order from the registry
    """
    return get_registry_record(WS_TEMPLATES_REGISTRY_NAME, default=[]) or []


def get_ordered_print_templates():
    """Returns an ordered list of print templates
    """
    order = get_print_template_order()
    templates = get_all_templates()
    ordered_templates = sorted(templates, key=lambda tpl: order.index(tpl) if tpl in order else len(order)))
    # filter out templates that are not contained in the order
    ordered_templates = filter(lambda tpl: tpl in order)
    return list(ordered_templates)

```

then simply return the results from `get_all_print_templates` from a `defaultFactory` here https://github.com/senaite/senaite.core/pull/2515/files#diff-079fe988789a9aae3125ed91e43ac1372530c685aa0f5031de7ffe121e0019b2R149

Reply:
oh, sorry, I didn’t understand right away, but I tried that and it doesn’t work :) - can not be calculating values for default registry fields

Reply:
sorry for long time answer and searching to rightly and suitable resolving

I think this is a universal solution that can be applied in the future